arangod/RocksDBEngine/RocksDBEdgeIndex.cpp

```diff
@@ -971,3 +1070,90 @@ void RocksDBEdgeIndex::recalculateEstimates() {
   }
   _estimator->setAppliedSeq(seq);
 }
+
+std::string_view RocksDBEdgeIndex::buildCompressedCacheKey(
+    std::string const*& previous, std::string_view value) const {
+  return _cacheKeyCollectionName.buildCompressedValue(previous, value);
+}
+
+std::string_view RocksDBEdgeIndex::buildCompressedCacheValue(
+    std::string const*& previous, std::string_view value) const {
+  return _cacheValueCollectionName.buildCompressedValue(previous, value);
+}
+
+std::string const* RocksDBEdgeIndex::cachedValueCollection(
+    std::string const*& previous) const noexcept {
+  if (previous == nullptr) {
+    previous = _cacheValueCollectionName.get();
+  }
+  TRI_ASSERT(previous != nullptr);
+  return previous;
+}
+
+RocksDBEdgeIndex::CachedCollectionName::CachedCollectionName() noexcept
+    : _name(nullptr) {}
+
+RocksDBEdgeIndex::CachedCollectionName::~CachedCollectionName() {
+  delete _name.load(std::memory_order_relaxed);
+}
+
+std::string_view RocksDBEdgeIndex::CachedCollectionName::buildCompressedValue(
+    std::string const*& previous, std::string_view value) const {
+  // split lookup value to determine collection name and key parts
+  auto pos = value.find('/');
+
+  if (pos == std::string_view::npos || pos == 0 || pos + 1 == value.size() ||
+      value[pos + 1] == '/') {
+    // totally invalid lookup value
+    value = {};
+  } else {
+    if (previous == nullptr) {
+      // no context yet. now try looking up cached collection name
+      previous = _name.load(std::memory_order_relaxed);
+      if (previous == nullptr) {
+        // no cached collection name yet. now try to store the collection name
+        // we determined ourselves. create a string with the collection name on
+        // the heap
+        auto cn = std::make_unique<std::string const>(value.data(), pos);
+        // try to store the name. this can race with other threads.
+        // TODO: fix memory order
+        if (_name.compare_exchange_strong(previous, cn.get())) {
+          // we won the race and were able to store our value. now we are owning
+          // the collection name.
+          previous = cn.release();
+        } else {
+          // we lost the race. now we need to fetch the name another thread
+          // stored.
+          // TODO: fix memory order
+          previous = _name.load();
+        }
+      }
+    }
+
+    // here we must have a collection name
+    TRI_ASSERT(previous != nullptr);
+
+    // now check if the collection name in the value we got matches the cached
+    // name.
+    TRI_ASSERT(!previous->empty());
+    // must have at least 'c/k'
+    TRI_ASSERT(value.size() > 2);
```

Comment:
You have code above which assign `{}` to value.

Reply:
Correct, but that assignment `value = {}` is in a different branch:
```
  if (pos == std::string_view::npos || pos == 0 || pos + 1 == value.size() ||
      value[pos + 1] == '/') {
    value = {};
```
Here we are in that `if`'s `else` block, so it does not matter.